File: src/AsciiMovie.Core/AmFrameLayout.cs
using System.Buffers.Binary;

namespace AsciiMovie.Core;

public static class AmFrameLayout
{
    public const ushort Version1 = 1;
    public const ushort Version2 = 2;

    public static int CharIndexSize(ushort version) => version >= Version2 ? 2 : 1;

    public static int CellStride(AmHeader header) =>
        CharIndexSize(header.Version) + (header.HasColor ? 3 : 0);

    public static int UncompressedFrameSize(AmHeader header) =>
        header.CellCount * CellStride(header);

    public static ushort GetCharIndex(ReadOnlySpan<byte> frame, int cellIndex, AmHeader header)
    {
        var offset = cellIndex * CellStride(header);
        return header.Version >= Version2
            ? BinaryPrimitives.ReadUInt16LittleEndian(frame.Slice(offset, 2))
            : frame[offset];
    }

    public static (byte R, byte G, byte B) GetColor(ReadOnlySpan<byte> frame, int cellIndex, AmHeader header)
    {
        var offset = cellIndex * CellStride(header) + CharIndexSize(header.Version);
        return (frame[offset], frame[offset + 1], frame[offset + 2]);
    }

    public static void SetCharIndex(Span<byte> frame, int cellIndex, AmHeader header, int charIndex)
    {
        var offset = cellIndex * CellStride(header);
        if (header.Version >= Version2)
            BinaryPrimitives.WriteUInt16LittleEndian(frame.Slice(offset, 2), (ushort)charIndex);
        else
            frame[offset] = (byte)charIndex;
    }

    public static void SetColor(Span<byte> frame, int cellIndex, AmHeader header, byte r, byte g, byte b)
    {
        var offset = cellIndex * CellStride(header) + CharIndexSize(header.Version);
        frame[offset] = r;
        frame[offset + 1] = g;
        frame[offset + 2] = b;
    }
}

File: web-plugin/src/AmFrameLayout.ts
export const AM_VERSION_1 = 1;
export const AM_VERSION_2 = 2;

export function charIndexSize(version: number): number {
  return version >= AM_VERSION_2 ? 2 : 1;
}

export function cellStride(header: { version: number; flags: number }, colorFlag: number): number {
  const color = (header.flags & colorFlag) !== 0;
  return charIndexSize(header.version) + (color ? 3 : 0);
}

export function uncompressedFrameSize(
  header: { version: number; flags: number; cols: number; rows: number },
  colorFlag: number,
): number {
  return header.cols * header.rows * cellStride(header, colorFlag);
}

export function getCharIndex(
  frame: Uint8Array,
  cellIndex: number,
  header: { version: number; flags: number },
  colorFlag: number,
): number {
  const offset = cellIndex * cellStride(header, colorFlag);
  if (header.version >= AM_VERSION_2) {
    return frame[offset] | (frame[offset + 1] << 8);
  }
  return frame[offset];
}

export function getColor(
  frame: Uint8Array,
  cellIndex: number,
  header: { version: number; flags: number },
  colorFlag: number,
): [number, number, number] {
  const offset = cellIndex * cellStride(header, colorFlag) + charIndexSize(header.version);
  return [frame[offset], frame[offset + 1], frame[offset + 2]];
}
